### PR TITLE
feat: set Component `bom_ref` to match `purl`

### PR DESCRIPTION
**Pull request by RodneyRichardson**
Partially addresses https://github.com/CycloneDX/cyclonedx-python/issues/355.

Add CLI argument to use purl.to_string() as bom_ref for Components. This should improve reproducibility.

### CODE REVIEW

**Review by jkowalleck — CHANGES_REQUESTED** [review]
@RodneyRichardson looks promissing.

a must change: 
could you raise the dependency _cyclonedx-python-lib_ from `>2.0.0` to the minimal version that supports the `bom-ref` kwarg on `Component.__init__()`? 
reason: the kwarg is not available in cyclonedx-python-lib@2.0 and will cause crashes.
see https://github.com/CycloneDX/cyclonedx-python-lib/blob/v2.0.0/cyclonedx/model/component.py

**Comment by jkowalleck** [comment]
https://github.com/CycloneDX/cyclonedx-python/pull/361#pullrequestreview-993082726
caused #363

**Comment by jkowalleck** [comment]
@RodneyRichardson could zou rebase on latest master branch,
so we can have the lowest-dependency-checks running (re: https://github.com/CycloneDX/cyclonedx-python/pull/361#issuecomment-1144539101)

**Comment by RodneyRichardson** [comment]
> a must change: could you raise the dependency _cyclonedx-python-lib_ from `>2.0.0` to the minimal version that supports the `bom-ref` kwarg on `Component.__init__()`? reason: the kwarg is not available in cyclonedx-python-lib@2.0 and will cause crashes. see https://github.com/CycloneDX/cyclonedx-python-lib/blob/v2.0.0/cyclonedx/model/component.py#L680

I'm not sure what you mean. The signature for `Component__init__()` in v2.0.0 contains `bom_ref`. Maybe the CI job will flag it up when they run.

**Comment by RodneyRichardson** [comment]
> @RodneyRichardson could you rebase on latest master branch, so we can have the lowest-dependency-checks running (re: [#361 (comment)](https://github.com/CycloneDX/cyclonedx-python/pull/361#issuecomment-1144539101))

I have updated from master. Is this sufficient?

**Comment by jkowalleck** [comment]
@RodneyRichardson 
you need to raise the dependency of _cyclonedx-python-lib_ to a version that allowes the setting of bom-ref in the constructor, still,

**Comment by jheck88** [comment]
I'd wait to merge this until there's a schema update to address this.  If you make this change you may impact existing automation that expects a uuid.  Let's get with Steve and see if there's some input.

**Comment by RodneyRichardson** [comment]
You may want to consider it in context with this other issue: https://github.com/CycloneDX/cyclonedx-python-lib/issues/239

**Comment by nscuro** [comment]
It should be noted that UUIDs can be generated in reproducible ways as well, by taking the contents of the described element into account. That'll be a fairly expensive operation though, so using PURL or similarly unique and human-readable properties should be preferred.

**Comment by jkowalleck** [comment]
to be honest the bom-ref's sole purpose is to be unique in the document's scope.
there was never a meaning intended for that string, but it could have one.

the core library is still free to use what ever bom-ref value you'd like, a uuid, a urn, a purl, ....
but the data generator in this tool here would benefit from a reproducible string - and purl would be candidate.

----

re: @jheck88 
> I'd wait to merge this until there's a schema update to address this. If you make this change you may impact existing automation that expects a uuid. Let's get with Steve and see if there's some input.

You are talking about this tool, not the https://github.com/CycloneDX/cyclonedx-python-lib ? 
You do have automation in pace, that expects the bom-ref of output generated with this exact tool to be uuis?
how come? Could you elaborate the constraints/requirements/environment?
bom-ref is defined as an arbitrary string per [spec](https://github.com/CycloneDX/specification/tree/master/schema), right?

**Comment by RodneyRichardson** [comment]
> It should be noted that UUIDs can be generated in reproducible ways as well, by taking the contents of the described element into account. That'll be a fairly expensive operation though, so using PURL or similarly unique and human-readable properties should be preferred.

That is true: something like this could work if we wanted reproducible UUIDs.
```
import hashlib
import uuid

m = hashlib.md5()
m.update(purl.to_string().encode('utf-8'))
new_uuid = uuid.UUID(m.hexdigest())
```

This could even go in `Component.__init__()` instead of generating a random UUID.

**Comment by jheck88** [comment]
> to be honest the bom-ref's sole purpose is to be unique in the document's scope. there was never a meaning intended for that string, but it could have one.
> 
> the core library is still free to use what ever bom-ref value you'd like, a uuid, a urn, a purl, .... but the data generator in this tool here would benefit from a reproducible string - and purl would be candidate.
> 
> re: @jheck88
> 
> > I'd wait to merge this until there's a schema update to address this. If you make this change you may impact existing automation that expects a uuid. Let's get with Steve and see if there's some input.
> 
> You are talking about this tool, not the https://github.com/CycloneDX/cyclonedx-python-lib ? You do have automation in pace, that expects the bom-ref of output generated with this exact tool to be uuis? how come? Could you elaborate the constraints/requirements/environment? bom-ref is defined as an arbitrary string per [spec](https://github.com/CycloneDX/specification/tree/master/schema), right?

More to the point that this field may deserve a regex or pattern in the schema.  If purl is correct then great, if a uuid generated off something like @RodneyRichardson comment above may work too.

**Comment by jkowalleck** [comment]
So, is there still a reason not to merge this feature? Everybody fine with this feature, @jheck88 ?

If not, we could add a CLI switch for this feature. @RodneyRichardson how do you feel about adding a switch for it?

**Comment by madpah** [comment]
Hi @jkowalleck , @RodneyRichardson , @jheck88

Please allow me to offer my opinion on this feature (apologies for the delay - I've been on vacation!)

**TL;DR:**
I only support this feature where it is explicitly enabled by a flag passed to `cyclonedx-python`.

**Long Read**
`bom_ref` is purposefully not a `purl` in our implementation here for a few reasons:

1. Not all components can be represented by a `purl`
2. The specification for CycloneDX includes a `bom_ref` and `purl` - there would be no reason to have both if `bom_ref` was best provided as a `purl`

So in short, from a "purist" view, I'm not supportive of updating `cyclonedx-python` or `cyclonedx-python-lib` to have `bom_ref` match `purl` automatically for Components. 

However, as `bom_ref` can be set (as identified in #355) in `cyclonedx-python` without changing the behaviour of the core `cyclonedx-python-library`, I would be happy to add a flag to `cyclonedx-python` that would optionally cause `cyclonedx-python` to use `purl` value as `bom_ref` where a `purl` exists on the Component.

**Review by madpah — CHANGES_REQUESTED** [review]
As per my comment on #361, I'm only supporting of this change at the current time when it is enabled by a feature flag.

**Comment by RodneyRichardson** [comment]
@madpah 
> 1. Not all components can be represented by a `purl`

It seems all components supported by cyclonedx-python can be identified with a purl (either pypi or conda). Can you give an example of one which cannot?

> 2. The specification for CycloneDX includes a `bom_ref` and `purl` - there would be no reason to have both if `bom_ref` was best provided as a `purl`

The guidance from CycloneDX is to use purl for the bom-ref, which is probably why a lot of the CycloneDX tools for other languages use it. From https://cyclonedx.org/use-cases/#dependency-graph: "Package URL (PURL) is an ideal choice for bom-ref as it will be both unique and readable. If PURL is not an option or not all components represented in the BOM contain a PURL, then UUID is recommended."

**Comment by RodneyRichardson** [comment]
> If not, we could add a CLI switch for this feature. @RodneyRichardson how do you feel about adding a switch for it?

I feel adding a switch is adding complexity that is not needed. I'd rather see the python tools move closer to creating reproducible output by default. 

If we did add a switch, I'd suggest `--random-bom-ref` to reinstate the old behaviour.

File: cyclonedx_py/parser/conda.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def _conda_packages_to_components(self) -> None:
         for conda_package in self._conda_packages:
             purl = conda_package_to_purl(conda_package)
             c = Component(
-                name=conda_package['name'], version=conda_package['version'],
+                name=conda_package['name'], bom_ref=purl.to_string(), version=str(conda_package['version']),
```
Comment (jkowalleck) [inline cyclonedx_py/parser/conda.py:67]: :question: 
could you explain, why the string cast was necessary? 
not that it is wrong, but i am worried whether there should be a regression test that failes when the string cast gets removed.

**Comment by jkowalleck** [comment]
> > If not, we could add a CLI switch for this feature. @RodneyRichardson how do you feel about adding a switch for it?
> 
> I feel adding a switch is adding complexity that is not needed. I'd rather see the python tools move closer to creating reproducible output by default.
> 
> If we did add a switch, I'd suggest `--random-bom-ref` to reinstate the old behaviour.
> 
> But, I can implement whichever way you want it to go - you are responsible for the direction of the tool.

I would suggest to implement a switch that enables this new feature. per default the bomRef should be randomized, to ensure uniqueness by most means.
The switch would enable bomRefs to be the same as the purl. 

we [learned](https://github.com/CycloneDX/cyclonedx-python/pull/361#issuecomment-1149331256)
from possible breaking changes if the default behavior changed.
I would be fine with these breaking changes. If you'd agree, then i would flag the original #355 as "breaking change" and you would remove the "fixes #355" from this PR so it does not close. 
This way we would merge your PR (get your feature in early), and could change the default behavior in the next major version.

File: cyclonedx_py/parser/conda.py (hunk 1 above)
Comment (RodneyRichardson) [reply cyclonedx_py/parser/conda.py:67]: I didn't add that - although it looks like I may have taken it out in a previous commit, and added it back in here! I'll take it out again.

Original code is here, which contains the cast, added by @madpah: https://github.com/CycloneDX/cyclonedx-python/commit/7bb6d328adec59cdd4c3ab80eb5f39568ca3bc9c#diff-97c8c273a746672bb6c4a9fa6e5595010c2b9b7efca301ae946946dc412af924

**Review by madpah — APPROVED** [review]
